### PR TITLE
Fix torch.where to accept tensors with different dtypes on CPU

### PR DESCRIPTION
**Pull request by ifedan**
Fix torch.where when called with arguments of different tensor types, as an example: https://github.com/pytorch/pytorch/issues/28709

we will calculate common dtype, based on two argument of torch.where, and use it for dispatching.

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -955,6 +955,36 @@ def test_where_bool_tensor(self):
...
+            else:
+                return torch.randint(-5, 5, size=size, dtype=dtype, device=device)
+
+        device = "cpu"
+        for dt1 in torch.testing.get_all_math_dtypes(device):
+            for dt2 in torch.testing.get_all_math_dtypes(device):
+                x1 = rand_tensor((5, 5), dt1, device)
+                x2 = rand_tensor((5, 5), dt2, device)
+                res = torch.where(x1 < 1, x1, x2)
+                print(dt1)
```
Comment (VitalyFedyunin) [inline test/test_torch.py]: No prints please

File: test/test_torch.py [hunk 2]
```diff
@@ -955,6 +955,36 @@ def test_where_bool_tensor(self):
...
+            if dtype.is_floating_point:
+                return torch.rand(size=size, dtype=dtype, device=device)
+            elif dtype == torch.uint8:
+                return torch.randint(1, 5, size=size, dtype=dtype, device=device)
+            elif dtype == torch.bool:
+                return torch.randint(0, 1, size=size, dtype=dtype, device=device).bool()
+            else:
+                return torch.randint(-5, 5, size=size, dtype=dtype, device=device)
+
+        device = "cpu"
```
Comment (VitalyFedyunin) [inline test/test_torch.py]: sure?

File: test/test_torch.py [hunk 3]
```diff
@@ -955,6 +955,36 @@ def test_where_bool_tensor(self):
...
+        for dt1 in torch.testing.get_all_math_dtypes(device):
+            for dt2 in torch.testing.get_all_math_dtypes(device):
+                x1 = rand_tensor((5, 5), dt1, device)
```
Comment (VitalyFedyunin) [inline test/test_torch.py]: nit: Please do something like h,w = 5, 5 and use this 'constants'

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]

**Comment by gchanan** [comment]
why do we have this logic only for CPU?  (actually, why do we have separate dispatch for where?).  Won't this have different behavior for CPU and CUDA otherwise?

File: test/test_torch.py [hunk 4]
```diff
@@ -955,6 +955,31 @@ def test_where_bool_tensor(self):
...
+                x2 = rand_tensor((5, 5), dt2, device)
+                res = torch.where(x1 < 1, x1, x2)
+                for i in range(5):
+                    for j in range(5):
+                        if x1[i][j] < 1:
+                            self.assertTrue(res[i][j].item() == x1[i][j].item())
```
Comment (gchanan) [inline test/test_torch.py]: non contiguous test please.

File: aten/src/ATen/native/TensorCompare.cpp [hunk 5]
```diff
@@ -122,9 +122,10 @@ std::vector<Tensor> where(const Tensor& condition) {
 }
 
 Tensor _s_where_cpu(const Tensor& condition, const Tensor& self, const Tensor& other) {
-  Tensor ret = at::empty(self.sizes(), self.options());
+  auto common_type = at::promote_types(self.scalar_type(), other.scalar_type());
```
Comment (gchanan) [inline aten/src/ATen/native/TensorCompare.cpp]: I believe this should be based on the result_type, not the promote_types.

Comment (nairbv) [reply aten/src/ATen/native/TensorCompare.cpp]: ^ yes we want to promote (using result_type) based on the tensor parameters, not directly on the dtypes. the tensors may be wrapped_numbers or zero-dim which gets handled differently.

File: test/test_torch.py [hunk 6]
```diff
@@ -955,6 +955,31 @@ def test_where_bool_tensor(self):
...
+            elif dtype == torch.uint8:
+                return torch.randint(1, 5, size=size, dtype=dtype, device=device)
+            elif dtype == torch.bool:
+                return torch.randint(0, 1, size=size, dtype=dtype, device=device).bool()
+            else:
+                return torch.randint(-5, 5, size=size, dtype=dtype, device=device)
+
+        device = "cpu"
+        for dt1 in torch.testing.get_all_math_dtypes(device):
+            for dt2 in torch.testing.get_all_math_dtypes(device):
```
Comment (nairbv) [inline test/test_torch.py]: nit: may want to use itertools.product to reduce indentation levels

File: test/test_torch.py [hunk 7]
```diff
@@ -955,6 +955,31 @@ def test_where_bool_tensor(self):
             res = torch.where(a > 0)
             self.assertEqual(1, len(res))
 
+    def test_where_tensor_cpu(self):
```
Comment (nairbv) [inline test/test_torch.py]: if we're mainly testing for the promotion cases, this might make more sense to include in test_type_promotion.py. We may also want to add `torch.where` to the lists of operators tested by some of the other existing promotion tests

File: aten/src/ATen/native/TensorCompare.cpp (hunk 5 above)
Comment (ifedan) [reply aten/src/ATen/native/TensorCompare.cpp]: Previously result type was calculating based on self dtype

Comment (nairbv) [reply aten/src/ATen/native/TensorCompare.cpp]: not sure what you mean by that. `at::tensor(self, other)` (from TypeProperties.cpp) accepts Tensors, promote_types only accepts dtypes.

Comment (ifedan) [reply aten/src/ATen/native/TensorCompare.cpp]: I see, make sense. I will change it to at::tensor(self, other) instead

Comment (nairbv) [reply aten/src/ATen/native/TensorCompare.cpp]: sorry I meant `result_type(self, other)`

Comment (ifedan) [reply aten/src/ATen/native/TensorCompare.cpp]: yes, I got it thanks

**Comment by ifedan** [comment]
This PR will be closed. Instead follow: https://github.com/pytorch/pytorch/pull/29078
